### PR TITLE
Improve Yarn workspaces handling

### PR DESCRIPTION
**Pull request by fviernau**
Fix several issues related to Yarn workspaces
    
    The existing logic for deciding between Yarn and NPM for a given list of
    packages was intentionally designed to be very basic, but it also exhibits
    several issues. As an incremental refactoring for fixing those seems to be
    much more effort this change re-writes that logic for mapping the definition
    files - that is the decision whether Yarn, Npm or none of them should handle
    a given 'package.json'.
    
    As of this change a definition file is handled by Yarn if and only if at
    least one of the folowing conditions is fulfilled:
    (1) it belongs to a Yarn workspace, e.g. either it is the root or a submodule
    (2) it has a yarn lockfile as sibling
    
    The logic for computing the root project directory was not working in some
    cases, e.g. given '/ab/package.json' and '/aa/package.json' it tries to
    locate the root project at '/a/package.json' which may not exist at all
    resulting in no workspace being found. Now the logic isn't based on
    'getCommonFilePrefix' which fixes that. Moreover this enables support
    for multiple Yarn workspaces as the detection is now independent of the
    actual project path.
    
    The logic for matching workspaces is extended to support the glob style
    wildcard matcher '*' and '**' for matching either a single or multiple
    directories. It is enabled by using the given glob pattern as is, but
    matching them against the project directory instead of against the
    definition file.
    
    Finally moving the code into a separate class elminates some knowledge
    from NPM about Yarn and more importantly makes NPM stateless again.

---
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/995)

### CODE REVIEW

File: analyzer/src/test/kotlin/PackageJsonUtilTest.kt [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+                """.trimIndent()
+                else "{}"
+
+        private fun mapDefinitionFiles(definitionFiles: Collection<File>) =
+                mapDefinitionFilesForNpm(definitionFiles).plus(mapDefinitionFilesForYarn(definitionFiles))
+    }
+
+    init {
+
+        "given no npm lockfile present, hasNpmLockFile returns false" {
```
Comment (mnonnenmacher) [inline analyzer/src/test/kotlin/PackageJsonUtilTest.kt]: Use "NPM" and "Yarn" instead of lowercase in the names of all test cases.

File: analyzer/src/test/kotlin/PackageJsonUtilTest.kt [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+            mapDefinitionFilesForYarn(definitionFiles) shouldContainExactly
+                    absolutePaths(
+                    "a/package.json")
+        }
+
+        "given project with no lockfile, definition file is mapped for npm only" {
+            setupProject(path = "a")
+
+            mapDefinitionFilesForNpm(definitionFiles) shouldContainExactly
+                    absolutePaths("a/package.json")
```
Comment (mnonnenmacher) [inline analyzer/src/test/kotlin/PackageJsonUtilTest.kt]: This fits in one line.

File: analyzer/src/test/kotlin/PackageJsonUtilTest.kt [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+            hasYarnLockFile("a") shouldBe true
+        }
+
+        "given project with yarn and npm lockfile, definition file is mapped for yarn only" {
+            setupProject(path = "a", hasNpmLockFile = true, hasYarnLockFile = true)
+
+            mapDefinitionFilesForNpm(definitionFiles).shouldBeEmpty()
+            mapDefinitionFilesForYarn(definitionFiles) shouldContainExactly
+                    absolutePaths(
+                    "a/package.json")
```
Comment (mnonnenmacher) [inline analyzer/src/test/kotlin/PackageJsonUtilTest.kt]: This fits in one line.

File: analyzer/src/test/kotlin/PackageJsonUtilTest.kt [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+            setupProject(path = "a", matchers = listOf("*", "*/f"))
+            setupProject(path = "a/b")
+            setupProject(path = "a/c")
+            setupProject(path = "a/d/e")
+            setupProject(path = "a/d/f")
+
+            mapDefinitionFiles(definitionFiles) shouldContainExactlyInAnyOrder
+                    absolutePaths(
+                            "a/package.json",
+                            "a/d/e/package.json")
```
Comment (mnonnenmacher) [inline analyzer/src/test/kotlin/PackageJsonUtilTest.kt]: When putting arguments in separate lines the closing paranthesis should go to next line according to the Kotlin coding conventions (same in the next test):
https://kotlinlang.org/docs/reference/coding-conventions.html#method-call-formatting

File: analyzer/src/test/kotlin/PackageJsonUtilTest.kt [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    override fun afterTest(description: Description, result: TestResult) {
+        tempDir.safeDeleteRecursively()
+        definitionFiles.clear()
+        super.afterTest(description, result)
+    }
+
+    private fun setupProject(path: String, matchers: List<String> = emptyList(),
+                             hasNpmLockFile: Boolean = false, hasYarnLockFile: Boolean = false) {
```
Comment (mnonnenmacher) [inline analyzer/src/test/kotlin/PackageJsonUtilTest.kt]: Use the syntax from the Kotlin coding conventions (yes, we don't do this consistently, but I'd like to change that):
https://kotlinlang.org/docs/reference/coding-conventions.html#function-formatting

File: analyzer/src/main/kotlin/managers/PackageJsonUtil.kt [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+internal class PackageJsonUtil {
+    companion object {
+        private val NPM_LOCK_FILES = listOf("npm-shrinkwrap.json", "package-lock.json")
+        private val YARN_LOCK_FILES = listOf("yarn.lock")
+
+        private data class DefinitionFileInfo(
+                val definitionFile: File,
+                var hasYarnLockfile: Boolean = false,
+                var hasNpmLockfile: Boolean = false,
+                var isYarnWorkspaceRoot: Boolean = false,
```
[the remark this answers is not in the document]
Comment (fviernau) [reply analyzer/src/main/kotlin/managers/PackageJsonUtil.kt]: I'd like to keep this as is because I think the  value of immutable is relatively low since the data class has very small visibility scope, while I think the current version is a bit more readable which seems more important to me. It would also keep out any logic out of the data class. Can we keep that as is?

File: analyzer/src/main/kotlin/managers/NPM.kt [hunk 7]
```diff
@@ -27,6 +27,7 @@ import com.fasterxml.jackson.databind.node.ObjectNode
 import com.here.ort.analyzer.AbstractPackageManagerFactory
 import com.here.ort.analyzer.HTTP_CACHE_PATH
 import com.here.ort.analyzer.PackageManager
+
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/NPM.kt]: Remove empty line.

File: analyzer/src/main/kotlin/managers/NPM.kt [hunk 8]
```diff
@@ -60,22 +60,18 @@ import java.net.HttpURLConnection
 import java.net.URI
 import java.net.URISyntaxException
 import java.net.URLEncoder
-import java.nio.file.FileSystems
-import java.nio.file.PathMatcher
-import java.util.SortedSet
+import java.util.*
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/NPM.kt]: No wildcard import for java.util please.

File: analyzer/src/main/kotlin/managers/NPM.kt [hunk 9]
```diff
@@ -60,22 +60,18 @@ import java.net.HttpURLConnection
...
 import org.apache.commons.codec.binary.Hex
 
-val NPM_LOCK_FILES = listOf("npm-shrinkwrap.json", "package-lock.json")
-val YARN_LOCK_FILES = listOf("yarn.lock")
-
 /**
  * The Node package manager for JavaScript, see https://www.npmjs.com/.
  */
 open class NPM(analyzerConfig: AnalyzerConfiguration, repoConfig: RepositoryConfiguration) :
+
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/NPM.kt]: Unrelated change.

File: analyzer/src/test/kotlin/PackageJsonUtilTest.kt [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+            hasNpmLockFile("a") shouldBe true
+        }
+
+        "given no yarn lockfile present, hasYarnLockFile returns false" {
+            setupProject(path = "a")
+
+            hasNpmLockFile("a") shouldBe false
+        }
+
+        "given yarn lockfile present, hasYarnLockFile returns true" {
```
Comment (mnonnenmacher) [inline analyzer/src/test/kotlin/PackageJsonUtilTest.kt]: Yarn

File: analyzer/src/test/kotlin/PackageJsonUtilTest.kt [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+            hasNpmLockFile("a") shouldBe false
+        }
+
+        "given NPM lockfile present, hasNpmLockFile returns true" {
+            setupProject(path = "a", hasNpmLockFile = true)
+
+            hasNpmLockFile("a") shouldBe true
+        }
+
+        "given no yarn lockfile present, hasYarnLockFile returns false" {
```
Comment (mnonnenmacher) [inline analyzer/src/test/kotlin/PackageJsonUtilTest.kt]: Yarn

File: analyzer/src/main/kotlin/managers/PackageJsonUtil.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply analyzer/src/main/kotlin/managers/PackageJsonUtil.kt]: Ok.
